### PR TITLE
Add documentation for widget manager & open handler

### PR DESCRIPTION
**Pull request by tortmayr**
- Document 'WidgetManager' and related interfaces
- Document 'WidgetOpenHandler' and related interfaces
Signed-off-by: Tobias Ortmayr <tortmayr@eclipsesource.com>
Contributed on behalf of STMicroelectronics

### CODE REVIEW

File: packages/core/src/browser/widget-manager.ts [hunk 1]
```diff
@@ -12,16 +12,40 @@
...
+ * export class MyWidget extends BaseWidget {
+ * }
+ *
+ * @injectable()
+ * export class MyWidgetFactory implements WidgetFactory {
+ *     id = 'myWidgetFactory';
+ *
+ *     createWidget(): MaybePromise<Widget> {
+ *         return new MyWidget();
+ *
```
Comment (planger) [inline packages/core/src/browser/widget-manager.ts:45]: This empty line seems unnecessary.
```suggestion
```

File: packages/core/src/browser/widget-open-handler.ts [hunk 2]
```diff
@@ -12,6 +12,10 @@
  * https://www.gnu.org/software/classpath/license.html.
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ *
+ * Contributors:
+ *     TypeFox- Initial API and implementation
```
Comment (planger) [inline packages/core/src/browser/widget-open-handler.ts:17]:
```suggestion
 *     TypeFox and others - Initial API and implementation
```

File: packages/core/src/browser/widget-open-handler.ts [hunk 3]
```diff
@@ -23,10 +27,13 @@ import { OpenHandler, OpenerOptions } from './opener-service';
 import { WidgetManager } from './widget-manager';
 
 export type WidgetOpenMode = 'open' | 'reveal' | 'activate';
-
+/**
+ * Representation of the {@link WidgetOpenerOptions}.
```
Comment (planger) [inline packages/core/src/browser/widget-open-handler.ts:31]:
```suggestion
```

File: packages/core/src/browser/widget-open-handler.ts [hunk 4]
```diff
@@ -23,10 +27,13 @@ import { OpenHandler, OpenerOptions } from './opener-service';
...
+ * Defines serializable generic options used by the {@link WidgetOpenHandler}.
```
Comment (planger) [inline packages/core/src/browser/widget-open-handler.ts:32]:
```suggestion
 * `WidgetOpenerOptions` define serializable generic options used by the {@link WidgetOpenHandler}.
```

File: packages/core/src/browser/widget-open-handler.ts [hunk 5]
```diff
@@ -37,6 +44,9 @@ export interface WidgetOpenerOptions extends OpenerOptions {
     widgetOptions?: ApplicationShell.WidgetOptions;
 }
 
+/**
+ * Generic base class for {@link Openhandler}s that are using a widget to open the given uri.
```
Comment (planger) [inline packages/core/src/browser/widget-open-handler.ts:48]:
```suggestion
 * Generic base class for {@link OpenHandler}s that are opening a widget for a given {@link URI}.
```

File: packages/core/src/browser/widget-open-handler.ts [hunk 6]
```diff
@@ -75,6 +85,10 @@ export abstract class WidgetOpenHandler<W extends BaseWidget> implements OpenHan
     /**
      * Open a widget for the given uri and options.
      * Reject if the given options is not an widget options or a widget cannot be opened.
```
Comment (planger) [inline packages/core/src/browser/widget-open-handler.ts:87]:
```suggestion
     * Reject if the given options are not widget options or a widget cannot be opened.
```

File: packages/core/src/browser/widget-open-handler.ts [hunk 7]
```diff
@@ -106,14 +123,19 @@ export abstract class WidgetOpenHandler<W extends BaseWidget> implements OpenHan
...
+     *
+     * @returns a promise of the existing or newly created widget.
      */
     getOrCreateByUri(uri: URI): Promise<W> {
         return this.getOrCreateWidget(uri);
     }
 
     /**
-     * All opened widgets.
+     * Retrieves all open widgets that haven been opened by this handler.
```
Comment (planger) [inline packages/core/src/browser/widget-open-handler.ts:136]:
```suggestion
     * Retrieves all open widgets that have been opened by this handler.
```

File: packages/core/src/browser/widget-open-handler.ts [hunk 8]
```diff
@@ -131,6 +153,12 @@ export abstract class WidgetOpenHandler<W extends BaseWidget> implements OpenHan
 
     protected abstract createWidgetOptions(uri: URI, options?: WidgetOpenerOptions): Object;
 
+    /**
+     * Closes all widgets that have been openend by this open handler.
+     * @param options the given close options that should be applied to all widgets.
```
Comment (planger) [inline packages/core/src/browser/widget-open-handler.ts:158]:
```suggestion
     * @param options the close options that should be applied to all widgets.
```

**Review by planger — COMMENTED** [review]
Excellent, thanks! I added a few suggestions inline!

**Comment by tortmayr** [comment]
Thanks. I applied your suggestions and squased the commits.

**Review by planger — APPROVED** [review]
LGTM, thanks a lot!

**Comment by sdirix** [comment]
Merged upstream with https://github.com/eclipse-theia/theia/pull/8644

**Pull request closed without merging**